File: Strings/reverse-words-in-a-string/reverse-words-in-a-string.go
// https://leetcode.com/problems/reverse-words-in-a-string/description/
package main

import "strings"
func reverseWords(s string) string {
    words := strings.Split(s, " ")
    // filter empty strings
    // filter undefined
    words = filter(words, func(word string) bool {
        return word != ""
    })
    for i, j := 0, len(words)-1; i < j; i, j = i+1, j-1 {
        words[i], words[j] = words[j], words[i]
    }
    return strings.Join(words, " ")
}


func filter(ss []string, test func(string) bool) (ret []string) {
    for _, s := range ss {
        if test(s) {
            ret = append(ret, s)
        }
    }
    return
}

File: Strings/reverse-words-in-a-string/reverse-words-in-a-string.rs
// https://leetcode.com/problems/reverse-words-in-a-string/description/

struct Solution;
impl Solution {
  pub fn reverse_words(s: String) -> String {
    s.split_whitespace()
      .rev()
      .collect::<Vec<&str>>()
      .join(" ")
  }
}

fn main() {
  let s = String::from("the sky is blue");
  let result = Solution::reverse_words(s);
  println!("{}", result);
}
